File: src/go/problems/tree_2_str_test.go
// https://leetcode.com/problems/construct-string-from-binary-tree

package problems

import (
	"strconv"
	"strings"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type Tree2str struct{}

func TestTree2str(t *testing.T) {
	gen := core.TestSuite[Tree2str]{}
	gen.Run(t)
}

func (Tree2str) Solution(root *TreeNode) string {
	var builder strings.Builder

	var build func(node *TreeNode)
	build = func(node *TreeNode) {
		builder.WriteString(strconv.Itoa(node.Val))

		if node.Left != nil {
			builder.WriteString("(")
			build(node.Left)
			builder.WriteString(")")
		}

		if node.Right != nil {
			if node.Left == nil {
				builder.WriteString("()")
			}

			builder.WriteString("(");
			build(node.Right);
			builder.WriteString(")");
		}
	}

	build(root)

	return builder.String()
}

File: src/python/problems/tree_2_str_test.py
# https://leetcode.com/problems/construct-string-from-binary-tree

from typing import Optional
from core.problem_base import *
from models.tree_node import TreeNode

class Tree2str(ProblemBase):
    def Solution(self, root: Optional[TreeNode]) -> str:
        result = ""

        def build(node):
            nonlocal result
            result += str(node.val)

            if node.left:
                result += "("
                build(node.left)
                result += ")"

            if node.right:
                if not node.left:
                    result += "()"

                result += "("
                build(node.right)
                result += ")"
        
        build(root)

        return result

if __name__ == '__main__':
    TestGen(Tree2str) \
        .Add(lambda tc: tc.ParamTreeNode([1,2,3,4]).Result("1(2(4))(3)")) \
        .Add(lambda tc: tc.ParamTreeNode([1,2,3,None,4]).Result("1(2()(4))(3)")) \
        .Run()
